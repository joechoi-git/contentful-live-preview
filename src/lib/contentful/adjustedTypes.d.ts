import type { IBlogPost, IBlogPostFields } from "./generatedTypes";

type Asset = {
    sys?: {
        id: string;
    };
    url: string;
};

type Document = {
    json: any;
};

export interface BlogDetailProps extends Omit<IBlogPostFields, "details" | "heroImage">, IBlogPost {
    sys: IBlogPost["sys"] & {
        publishedAt: string;
        publishedVersion: number;
    };
    details: Document;
    heroImage: Asset;
    relatedBlogsCollection?: {
        items: BlogDetailProps[];
    };
}
