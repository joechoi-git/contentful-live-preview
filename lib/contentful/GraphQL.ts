import { DEFAULT_LANGUAGE } from "./Constants";
import { convertLocale } from "./Utils";

/* eslint-disable indent */
const BLOG_GRAPHQL_FIELDS = `
    sys {
        id
        publishedAt
        publishedVersion
    }
    __typename
    title
    slug
    summary
    heroImage {
        sys {
            id
        }
        __typename
        url
    }
    date
    author
    categoryName
    carousel
    details {
        json
    }
    relatedBlogsCollection {
        items {
            sys {
                id
                publishedAt
                publishedVersion
            }
            title
            summary
            heroImage {
                sys {
                    id
                }
                __typename
                url
            }
            slug
            date
            author
            categoryName
        }
    }
`;

// Set a variable that contains all the fields needed for blogs when a fetch for content is performed
/*
// GraphQL playground
query {
  blogPostCollection(where: { slug: "supporting-neurodevelopment-of-pediatric-heart-patients-4-new-initiatives" }) {
    items {
      sys {
        id
      }
      title
      summary
      heroImage {
        url
      }
      slug
      date
      author
      categoryName
      carousel
      details {
        json
      }
      relatedBlogsCollection {
        items {
          sys {
            id
          }
          title
          summary
          heroImage {
            url
          }
          slug
          date
          author
          categoryName
        }
      }
    }
  }
}
*/

async function fetchGraphQL(query: string, preview = false, tags: [string] = [""]) {
    return fetch(
        `https://graphql.contentful.com/content/v1/spaces/${process.env.NEXT_PUBLIC_CONTENTFUL_SPACE_ID}/environments/${process.env.NEXT_PUBLIC_CONTENTFUL_ENVIRONMENT}`,
        {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                // Switch the Bearer token depending on whether the fetch is supposed to retrieve draft or published content
                Authorization: `Bearer ${
                    preview
                        ? process.env.CONTENTFUL_PREVIEW_ACCESS_TOKEN
                        : process.env.CONTENTFUL_ACCESS_TOKEN
                }`
            },
            body: JSON.stringify({ query }),
            // Add a Next.js specific header with tags to revalidate the content:
            // Associate all fetches for blogs with an "blogs" cache tag so content can be revalidated or updated from Contentful on publish
            next: { tags, revalidate: 0 }
        }
    ).then((response) => response.json());
}

function extractBlogEntries(fetchResponse: { data: { blogPostCollection: { items: any } } }) {
    return fetchResponse?.data?.blogPostCollection?.items;
}

export async function getAllBlogs(limit = 20, isDraftMode = false, locale = DEFAULT_LANGUAGE) {
    // console.log("getAllBlogs", limit, isDraftMode, locale);
    const blogs = await fetchGraphQL(
        `query {
      blogPostCollection(where:{slug_exists: true}, locale: "${convertLocale(locale)}", order: date_DESC, limit: ${limit}, preview: ${
          isDraftMode ? "true" : "false"
      }) {
          items {
            ${BLOG_GRAPHQL_FIELDS}
          }
        }
      }`,
        isDraftMode,
        ["blogs"]
    );
    // console.log("getAllBlogs", JSON.stringify(blogs, null, 4));
    return extractBlogEntries(blogs);
}

export async function getBlog(slug: string, isDraftMode = false, locale = DEFAULT_LANGUAGE) {
    // console.log("getBlog", slug, isDraftMode, locale);
    const blog = await fetchGraphQL(
        `query {
      blogPostCollection(where:{slug: "${slug}"}, locale: "${convertLocale(locale)}", limit: 1, preview: ${
          isDraftMode ? "true" : "false"
      }) {
          items {
            ${BLOG_GRAPHQL_FIELDS}
          }
        }
      }`,
        isDraftMode,
        [slug]
    );
    // console.log("getBlog", JSON.stringify(blog, null, 4));
    return extractBlogEntries(blog)[0];
}
