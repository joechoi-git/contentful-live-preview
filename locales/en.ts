export default {
    hello: "Hello",
    "hello.world": "Hello world!",
    welcome: "Hello {name}!",
    home: "Home",
    article: "Article",
    researchLibrary: "Research Library",
    exploreLibrary: "Explore the latest innovations, clinical trials and research articles"
} as const;
