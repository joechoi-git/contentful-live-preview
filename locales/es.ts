export default {
    hello: "Hola",
    "hello.world": "Hola world!",
    welcome: "Hola {name}!",
    home: "Hogar",
    article: "Artículo",
    researchLibrary: "Biblioteca de investigación",
    exploreLibrary:
        "Explore las últimas innovaciones, ensayos clínicos y artículos de investigación."
} as const;
