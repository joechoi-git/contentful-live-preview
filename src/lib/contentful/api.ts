/* eslint-disable indent */
const BLOG_GRAPHQL_FIELDS = `
    sys {
        id
        publishedAt
        publishedVersion
    }
    __typename
    title
    slug
    summary
    details {
        json
    }
    date
    author
    categoryName
    heroImage {
        sys {
            id
        }
        __typename
        url
    }
    relatedBlogsCollection {
        items {
            sys {
                id
                publishedAt
                publishedVersion
            }
            title
            summary
            heroImage {
                sys {
                    id
                }
                __typename
                url
            }
            slug
            date
            author
            categoryName
        }
    }
`;

// Set a variable that contains all the fields needed for blogs when a fetch for content is performed
/*
// GraphQL playground
query {
  blogPostCollection(where: { slug: "supporting-neurodevelopment-of-pediatric-heart-patients-4-new-initiatives" }) {
    items {
      sys {
        id
      }
      title
      summary
      heroImage {
        url
      }
      slug
      details {
        json
      }
      date
      author
      categoryName
      relatedBlogsCollection {
        items {
          sys {
            id
          }
          title
          summary
          heroImage {
            url
          }
          slug
          date
          author
          categoryName
        }
      }
    }
  }
}
*/

async function fetchGraphQL(query: string, preview = false, tags: [string] = [""]) {
    return fetch(
        `https://graphql.contentful.com/content/v1/spaces/${process.env.NEXT_PUBLIC_CONTENTFUL_SPACE_ID}/environments/${process.env.NEXT_PUBLIC_CONTENTFUL_ENVIRONMENT}`,
        {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                // Switch the Bearer token depending on whether the fetch is supposed to retrieve draft or published content
                Authorization: `Bearer ${
                    preview
                        ? process.env.CONTENTFUL_PREVIEW_ACCESS_TOKEN
                        : process.env.CONTENTFUL_ACCESS_TOKEN
                }`
            },
            body: JSON.stringify({ query }),
            // Add a Next.js specific header with tags to revalidate the content:
            // Associate all fetches for blogs with an "blogs" cache tag so content can be revalidated or updated from Contentful on publish
            next: { tags, revalidate: 0 }
        }
    ).then((response) => response.json());
}

function extractBlogEntries(fetchResponse: { data: { blogPostCollection: { items: any } } }) {
    return fetchResponse?.data?.blogPostCollection?.items;
}

export async function getAllBlogs(limit = 20, isDraftMode = false) {
    const blogs = await fetchGraphQL(
        `query {
      blogPostCollection(where:{slug_exists: true}, order: date_DESC, limit: ${limit}, preview: ${
          isDraftMode ? "true" : "false"
      }) {
          items {
            ${BLOG_GRAPHQL_FIELDS}
          }
        }
      }`,
        isDraftMode,
        ["blogs"]
    );

    return extractBlogEntries(blogs);
}

export async function getBlog(slug: string, isDraftMode = false) {
    const blog = await fetchGraphQL(
        `query {
      blogPostCollection(where:{slug: "${slug}"}, limit: 1, preview: ${
          isDraftMode ? "true" : "false"
      }) {
          items {
            ${BLOG_GRAPHQL_FIELDS}
          }
        }
      }`,
        isDraftMode,
        [slug]
    );
    return extractBlogEntries(blog)[0];
}
