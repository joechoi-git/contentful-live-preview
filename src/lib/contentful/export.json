{
  "contentTypes": [
    {
      "sys": {
        "space": {
          "sys": {
            "type": "Link",
            "linkType": "Space",
            "id": "2asdi3p6ywuq"
          }
        },
        "id": "blogPost",
        "type": "ContentType",
        "createdAt": "2024-06-10T18:11:13.334Z",
        "updatedAt": "2024-07-24T18:07:03.017Z",
        "environment": {
          "sys": {
            "id": "master",
            "type": "Link",
            "linkType": "Environment"
          }
        },
        "publishedVersion": 19,
        "publishedAt": "2024-07-24T18:07:03.017Z",
        "firstPublishedAt": "2024-06-10T18:11:13.693Z",
        "createdBy": {
          "sys": {
            "type": "Link",
            "linkType": "User",
            "id": "1ZQUn59WtWWExkKm2Tw3qe"
          }
        },
        "updatedBy": {
          "sys": {
            "type": "Link",
            "linkType": "User",
            "id": "1ZQUn59WtWWExkKm2Tw3qe"
          }
        },
        "publishedCounter": 10,
        "version": 20,
        "publishedBy": {
          "sys": {
            "type": "Link",
            "linkType": "User",
            "id": "1ZQUn59WtWWExkKm2Tw3qe"
          }
        },
        "urn": "crn:contentful:::content:spaces/2asdi3p6ywuq/environments/master/content_types/blogPost"
      },
      "displayField": "title",
      "name": "Blog Post",
      "description": "",
      "fields": [
        {
          "id": "slug",
          "name": "Slug",
          "type": "Symbol",
          "localized": false,
          "required": true,
          "validations": [
            {
              "unique": true
            }
          ],
          "disabled": false,
          "omitted": false
        },
        {
          "id": "title",
          "name": "Title",
          "type": "Symbol",
          "localized": false,
          "required": true,
          "validations": [
          ],
          "disabled": false,
          "omitted": false
        },
        {
          "id": "summary",
          "name": "Summary",
          "type": "Symbol",
          "localized": false,
          "required": true,
          "validations": [
          ],
          "disabled": false,
          "omitted": false
        },
        {
          "id": "heroImage",
          "name": "Hero Image",
          "type": "Link",
          "localized": false,
          "required": true,
          "validations": [
          ],
          "disabled": false,
          "omitted": false,
          "linkType": "Asset"
        },
        {
          "id": "date",
          "name": "Date",
          "type": "Date",
          "localized": false,
          "required": true,
          "validations": [
          ],
          "disabled": false,
          "omitted": false
        },
        {
          "id": "author",
          "name": "Author",
          "type": "Symbol",
          "localized": false,
          "required": true,
          "validations": [
          ],
          "disabled": false,
          "omitted": false
        },
        {
          "id": "categoryName",
          "name": "Category",
          "type": "Symbol",
          "localized": false,
          "required": true,
          "validations": [
          ],
          "disabled": false,
          "omitted": false
        },
        {
          "id": "carousel",
          "name": "Carousel",
          "type": "Object",
          "localized": false,
          "required": false,
          "validations": [
          ],
          "disabled": false,
          "omitted": false
        },
        {
          "id": "details",
          "name": "Details",
          "type": "RichText",
          "localized": false,
          "required": true,
          "validations": [
            {
              "enabledMarks": [
                "bold",
                "italic",
                "underline",
                "code",
                "superscript",
                "subscript"
              ],
              "message": "Only bold, italic, underline, code, superscript, and subscript marks are allowed"
            },
            {
              "enabledNodeTypes": [
                "heading-1",
                "heading-2",
                "heading-3",
                "heading-4",
                "heading-5",
                "heading-6",
                "unordered-list",
                "ordered-list",
                "blockquote",
                "hr",
                "table",
                "hyperlink",
                "entry-hyperlink",
                "asset-hyperlink",
                "embedded-entry-block",
                "embedded-entry-inline",
                "embedded-asset-block"
              ],
              "message": "Only heading 1, heading 2, heading 3, heading 4, heading 5, heading 6, unordered list, ordered list, quote, horizontal rule, table, link to Url, link to entry, link to asset, block entry, inline entry, and asset nodes are allowed"
            },
            {
              "nodes": {
              }
            }
          ],
          "disabled": false,
          "omitted": false
        },
        {
          "id": "relatedBlogs",
          "name": "Related Blogs",
          "type": "Array",
          "localized": false,
          "required": false,
          "validations": [
            {
              "size": {
                "min": 0,
                "max": 5
              }
            }
          ],
          "disabled": false,
          "omitted": false,
          "items": {
            "type": "Link",
            "validations": [
              {
                "linkContentType": [
                  "blogPost"
                ]
              }
            ],
            "linkType": "Entry"
          }
        }
      ],
      "metadata": {
        "annotations": {
          "ContentType": [
            {
              "sys": {
                "id": "Contentful:AggregateRoot",
                "type": "Link",
                "linkType": "Annotation"
              }
            }
          ]
        }
      }
    }
  ],
  "tags": [
  ],
  "editorInterfaces": [
    {
      "sys": {
        "id": "default",
        "type": "EditorInterface",
        "space": {
          "sys": {
            "id": "2asdi3p6ywuq",
            "type": "Link",
            "linkType": "Space"
          }
        },
        "version": 18,
        "createdAt": "2024-06-10T18:11:13.883Z",
        "createdBy": {
          "sys": {
            "id": "1ZQUn59WtWWExkKm2Tw3qe",
            "type": "Link",
            "linkType": "User"
          }
        },
        "updatedAt": "2024-07-24T18:07:03.599Z",
        "updatedBy": {
          "sys": {
            "id": "1ZQUn59WtWWExkKm2Tw3qe",
            "type": "Link",
            "linkType": "User"
          }
        },
        "contentType": {
          "sys": {
            "id": "blogPost",
            "type": "Link",
            "linkType": "ContentType"
          }
        },
        "environment": {
          "sys": {
            "id": "master",
            "type": "Link",
            "linkType": "Environment"
          }
        }
      },
      "editors": [
        {
          "settings": {
          },
          "widgetId": "default-editor",
          "widgetNamespace": "editor-builtin"
        },
        {
          "settings": {
          },
          "widgetId": "tags-editor",
          "widgetNamespace": "editor-builtin"
        },
        {
          "settings": {
          },
          "widgetId": "3fLf4DnjbkN1yUA2V2fMgE",
          "widgetNamespace": "app"
        }
      ],
      "controls": [
        {
          "fieldId": "slug",
          "widgetId": "singleLine",
          "widgetNamespace": "builtin"
        },
        {
          "fieldId": "title",
          "widgetId": "singleLine",
          "widgetNamespace": "builtin"
        },
        {
          "fieldId": "summary",
          "widgetId": "singleLine",
          "widgetNamespace": "builtin"
        },
        {
          "fieldId": "heroImage",
          "widgetId": "assetLinkEditor",
          "widgetNamespace": "builtin"
        },
        {
          "fieldId": "date",
          "widgetId": "datePicker",
          "widgetNamespace": "builtin"
        },
        {
          "fieldId": "author",
          "widgetId": "singleLine",
          "widgetNamespace": "builtin"
        },
        {
          "fieldId": "categoryName",
          "widgetId": "singleLine",
          "widgetNamespace": "builtin"
        },
        {
          "fieldId": "carousel",
          "widgetId": "5KySdUzG7OWuCE2V3fgtIa",
          "widgetNamespace": "app"
        },
        {
          "fieldId": "details",
          "widgetId": "richTextEditor",
          "widgetNamespace": "builtin"
        },
        {
          "fieldId": "relatedBlogs",
          "settings": {
            "bulkEditing": false,
            "showLinkEntityAction": true,
            "showCreateEntityAction": false
          },
          "widgetId": "entryLinksEditor",
          "widgetNamespace": "builtin"
        }
      ],
      "editorLayout": [
        {
          "name": "Content",
          "items": [
            {
              "fieldId": "relatedBlogs"
            },
            {
              "fieldId": "carousel"
            }
          ],
          "groupId": "content"
        },
        {
          "name": "Page Settings",
          "items": [
            {
              "fieldId": "title"
            },
            {
              "fieldId": "slug"
            },
            {
              "fieldId": "summary"
            },
            {
              "fieldId": "details"
            },
            {
              "fieldId": "date"
            },
            {
              "fieldId": "author"
            },
            {
              "fieldId": "categoryName"
            },
            {
              "fieldId": "heroImage"
            }
          ],
          "groupId": "pageSettings"
        }
      ],
      "groupControls": [
        {
          "groupId": "content",
          "widgetId": "topLevelTab",
          "widgetNamespace": "builtin"
        },
        {
          "groupId": "pageSettings",
          "widgetId": "topLevelTab",
          "widgetNamespace": "builtin"
        }
      ]
    }
  ],
  "locales": [
    {
      "name": "English (United States)",
      "code": "en-US",
      "fallbackCode": null,
      "default": true,
      "contentManagementApi": true,
      "contentDeliveryApi": true,
      "optional": false,
      "sys": {
        "type": "Locale",
        "id": "7evhqDuDCw69zzXDU1SHvq",
        "version": 1,
        "space": {
          "sys": {
            "type": "Link",
            "linkType": "Space",
            "id": "2asdi3p6ywuq"
          }
        },
        "environment": {
          "sys": {
            "type": "Link",
            "linkType": "Environment",
            "id": "master-2024-07-22",
            "uuid": "3013a626-5629-4026-9e4a-a0510856327c"
          }
        },
        "createdBy": {
          "sys": {
            "type": "Link",
            "linkType": "User",
            "id": "1ZQUn59WtWWExkKm2Tw3qe"
          }
        },
        "createdAt": "2024-06-10T18:05:22Z",
        "updatedBy": {
          "sys": {
            "type": "Link",
            "linkType": "User",
            "id": "1ZQUn59WtWWExkKm2Tw3qe"
          }
        },
        "updatedAt": "2024-06-10T18:05:22Z"
      }
    }
  ],
  "webhooks": [
    {
      "name": "Revalidate Webhook",
      "url": "https://contentful-live-preview-two.vercel.app/api/revalidate",
      "httpBasicUsername": null,
      "topics": [
        "Entry.*"
      ],
      "filters": null,
      "transformation": {
        "contentType": "application/json",
        "body": {
          "tag": "{ /payload/fields/slug/en-US }"
        }
      },
      "active": true,
      "sys": {
        "type": "WebhookDefinition",
        "id": "43zR4Bt7hqCbUW1pasBfLB",
        "version": 1,
        "space": {
          "sys": {
            "type": "Link",
            "linkType": "Space",
            "id": "2asdi3p6ywuq"
          }
        },
        "createdBy": {
          "sys": {
            "type": "Link",
            "linkType": "User",
            "id": "1ZQUn59WtWWExkKm2Tw3qe"
          }
        },
        "createdAt": "2024-06-10T18:23:55Z",
        "updatedBy": {
          "sys": {
            "type": "Link",
            "linkType": "User",
            "id": "1ZQUn59WtWWExkKm2Tw3qe"
          }
        },
        "updatedAt": "2024-06-10T18:45:48Z"
      },
      "headers": [
        {
          "key": "x-vercel-reval-key",
          "secret": true
        }
      ]
    }
  ],
  "roles": [
    {
      "name": "Editor",
      "description": "Allows editing, publishing and archiving of content",
      "policies": [
        {
          "effect": "allow",
          "actions": "all",
          "constraint": {
            "and": [
              {
                "equals": [
                  {
                    "doc": "sys.type"
                  },
                  "Entry"
                ]
              }
            ]
          }
        },
        {
          "effect": "allow",
          "actions": "all",
          "constraint": {
            "and": [
              {
                "equals": [
                  {
                    "doc": "sys.type"
                  },
                  "Asset"
                ]
              }
            ]
          }
        }
      ],
      "permissions": {
        "ContentModel": [
          "read"
        ],
        "Settings": [
        ],
        "ContentDelivery": [
        ],
        "Environments": [
        ],
        "EnvironmentAliases": [
        ],
        "Tags": [
        ]
      },
      "sys": {
        "type": "Role",
        "id": "7eLSEhYphfVLeMK1pdKgxG",
        "version": 0,
        "space": {
          "sys": {
            "type": "Link",
            "linkType": "Space",
            "id": "2asdi3p6ywuq"
          }
        },
        "createdBy": {
          "sys": {
            "type": "Link",
            "linkType": "User",
            "id": "1ZQUn59WtWWExkKm2Tw3qe"
          }
        },
        "createdAt": "2024-06-10T18:05:22Z",
        "updatedBy": {
          "sys": {
            "type": "Link",
            "linkType": "User",
            "id": "1ZQUn59WtWWExkKm2Tw3qe"
          }
        },
        "updatedAt": "2024-06-10T18:05:22Z"
      }
    }
  ]
}