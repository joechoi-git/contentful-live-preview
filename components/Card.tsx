import React from "react";
import Link from "next/link";
import Image from "next/image";
import type { BlogDetail, RelatedBlog } from "@/lib/contentful/api"; // Adjust the import path as necessary
import { isoToFriendlyDate, isoToFriendlyDateTime } from "@/lib/contentful/options"; // Adjust the import path as necessary

interface CardProps {
    blog: RelatedBlog | BlogDetail;
}

const Card: React.FC<CardProps> = ({ blog }) => {
    return blog && blog.slug ? (
        <article className="h-full flex flex-col rounded-lg shadow-lg overflow-hidden">
            <Link href={`/blogs/${blog.slug}`}>
                <Image
                    alt="placeholder"
                    className="aspect-[4/3] object-cover w-full"
                    height="263"
                    src={blog?.heroImage?.url ?? "/default.jpeg"}
                    width="350"
                />
            </Link>
            <div className="flex-1 p-6">
                <Link href={`/blogs/${blog.slug}`}>
                    <h3 className="text-2xl font-bold leading-tight text-zinc-900 dark:text-zinc-50 py-4">
                        {blog.title}
                    </h3>
                </Link>
                <div className="inline-block rounded-full bg-zinc-100 px-3 py-1 text-sm font-semibold text-zinc-800">
                    {blog.categoryName}
                </div>
                <p className="max-w-none text-zinc-500 mt-4 mb-2 text-sm dark:text-zinc-400">
                    {blog.summary}
                </p>
                <p className="max-w-none text-zinc-600 mt-2 mb-2 text-sm font-bold dark:text-zinc-400">
                    Authored By: {blog.author}
                </p>
                <p className="max-w-none text-zinc-600 mt-2 mb-2 text-sm font-bold dark:text-zinc-400">
                    Authored On: {blog.date ? isoToFriendlyDate(blog.date.toString()) : ""}
                </p>
                <p className="max-w-none text-red-600 mt-2 mb-2 text-sm font-bold dark:text-red-400">
                    Published On:{" "}
                    {blog.sys.publishedAt
                        ? isoToFriendlyDateTime(blog.sys.publishedAt.toString())
                        : "In Draft"}
                </p>
                <div className="flex justify-end">
                    <Link
                        className="inline-flex h-10 items-center justify-center text-sm font-medium"
                        href={`/blogs/${blog.slug}`}
                    >
                        Read More â†’
                    </Link>
                </div>
            </div>
        </article>
    ) : null;
};

export default Card;
