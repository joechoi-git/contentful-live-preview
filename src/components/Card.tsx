import React from "react";
import Link from "next/link";
import Image from "next/image";
import type { BlogDetailProps } from "../lib/contentful/adjustedTypes";
import { isoToFriendlyDate, isoToFriendlyDateTime } from "../lib/contentful/Utils";

interface CardProps {
    blog: BlogDetailProps;
}

const Card: React.FC<CardProps> = ({ blog }) => {
    return blog && blog.slug ? (
        <article className="h-full flex flex-col border-2 border-secondary rounded-xl shadow-xl overflow-hidden">
            <Link href={`/blogs/${blog.slug}`}>
                <Image
                    alt="placeholder"
                    className={`aspect-[4/3] object-cover w-full ${blog.sys.publishedAt ? "" : "grayscale"}`}
                    height="263"
                    src={blog?.heroImage?.url ?? "/default.jpeg"}
                    width="350"
                />
            </Link>
            <div className={`flex-1 p-6 ${blog.sys.publishedAt ? "bg-base-100" : "bg-base-300"}`}>
                <Link href={`/blogs/${blog.slug}`}>
                    <h3 className="text-2xl font-bold leading-tight py-4 text-accent">
                        {blog.title}
                    </h3>
                </Link>
                <div className="font-bold">{blog.categoryName}</div>
                <p className="mt-4 mb-2">{blog.summary}</p>
                <p className="mt-2 mb-2 font-bold">Authored By: {blog.author}</p>
                <p className="mt-2 mb-2 font-bold">
                    Authored On: {blog.date ? isoToFriendlyDate(blog.date.toString()) : ""}
                </p>
                <p className="mt-2 mb-2 font-bold text-secondary">
                    Published On:{" "}
                    {blog.sys.publishedAt
                        ? isoToFriendlyDateTime(blog.sys.publishedAt.toString())
                        : "In Draft"}
                </p>
                <div className="flex justify-end">
                    <Link
                        className="inline-flex h-10 items-center justify-center font-medium text-accent"
                        href={`/blogs/${blog.slug}`}
                    >
                        Read More â†’
                    </Link>
                </div>
            </div>
        </article>
    ) : null;
};

export default Card;
