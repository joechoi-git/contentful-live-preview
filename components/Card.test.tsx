import React from "react";
import { render, screen } from "@testing-library/react";
import Card from "./Card";
import { useI18n } from "../locales/client";
import { CardProps } from "../lib/contentful/adjustedTypes";

// Mock the `useI18n` hook
jest.mock("../locales/client", () => ({
    useI18n: jest.fn()
}));

describe("Card component", () => {
    const mockBlog: CardProps = {
        slug: "test-blog",
        title: "Test Blog Title",
        categoryName: "Category Name",
        summary: "This is a summary of the blog post.",
        author: "Author Name",
        date: new Date().toISOString(),
        heroImage: {
            url: "/image.jpg"
            // Add any other properties required by the Asset type
        },
        sys: {
            id: "unique-id",
            type: "Entry",
            createdAt: new Date().toISOString(),
            updatedAt: new Date().toISOString(),
            locale: "en-US",
            contentType: {
                sys: {
                    id: "blogPost",
                    linkType: "ContentType",
                    type: "Link"
                }
            },
            publishedAt: "", // Use an empty string or another placeholder instead of null
            publishedVersion: 1
        },
        metadata: {
            tags: []
        }, // Add appropriate mock data here if needed
        fields: {} // Add appropriate mock data here if needed
    };

    beforeEach(() => {
        (useI18n as jest.Mock).mockReturnValue((key: string) => key);
    });

    test("renders the Card component with provided blog data", () => {
        render(<Card blog={mockBlog} />);

        // Check if the title is rendered
        expect(screen.getByText("Test Blog Title")).toBeInTheDocument();

        // Check if the category name is rendered
        expect(screen.getByText("Category Name")).toBeInTheDocument();

        // Check if the summary is rendered
        expect(screen.getByText("This is a summary of the blog post.")).toBeInTheDocument();

        // Check if the author is rendered
        expect(screen.getByText("authoredBy: Author Name")).toBeInTheDocument();

        // Check if the date is rendered
        expect(screen.getByText(/authoredOn:/)).toBeInTheDocument();

        // Check if the "read more" link is rendered
        expect(screen.getByText("readMore â†’")).toBeInTheDocument();
    });

    test("renders 'In Draft' if the published date is not provided", () => {
        const draftBlog = {
            ...mockBlog,
            sys: {
                ...mockBlog.sys,
                publishedAt: "" // Use an empty string instead of null
            }
        };
        render(<Card blog={draftBlog} />);

        expect(screen.getByText("publishedOn: In Draft")).toBeInTheDocument();
    });

    test("does not render the component if the blog has no slug", () => {
        const invalidBlog = { ...mockBlog, slug: "" };
        const { container } = render(<Card blog={invalidBlog} />);
        expect(container.firstChild).toBeNull();
    });

    test("matches snapshot", () => {
        const { asFragment } = render(<Card blog={mockBlog} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
