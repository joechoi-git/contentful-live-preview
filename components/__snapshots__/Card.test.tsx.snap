// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component matches snapshot 1`] = `
<DocumentFragment>
  <article
    class="h-full flex flex-col border-2 border-secondary rounded-xl shadow-xl overflow-hidden"
  >
    <a
      href="/blogs/test-blog"
    >
      <img
        alt="placeholder"
        class="aspect-[4/3] object-cover w-full grayscale"
        data-nimg="1"
        decoding="async"
        height="263"
        loading="lazy"
        src="/_next/image?url=%2Fimage.jpg&w=750&q=75"
        srcset="/_next/image?url=%2Fimage.jpg&w=384&q=75 1x, /_next/image?url=%2Fimage.jpg&w=750&q=75 2x"
        style="color: transparent;"
        width="350"
      />
    </a>
    <div
      class="flex-1 p-6 bg-base-300"
    >
      <a
        href="/blogs/test-blog"
      >
        <h3
          class="text-2xl font-bold leading-tight py-4 text-accent"
        >
          Test Blog Title
        </h3>
      </a>
      <div
        class="font-bold"
      >
        Category Name
      </div>
      <p
        class="mt-4 mb-2"
      >
        This is a summary of the blog post.
      </p>
      <p
        class="mt-2 mb-2 font-bold"
      >
        authoredBy: Author Name
      </p>
      <p
        class="mt-2 mb-2 font-bold"
      >
        authoredOn: September 4, 2024
      </p>
      <p
        class="mt-2 mb-2 font-bold text-secondary"
      >
        publishedOn: In Draft
      </p>
      <div
        class="flex justify-end"
      >
        <a
          class="inline-flex h-10 items-center justify-center font-medium text-accent"
          href="/blogs/test-blog"
        >
          readMore â†’
        </a>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
